c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -244,47 +263,71 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
         // the rewrite should occur at the callsite
         let callsite_span = ex.span.source_callsite();
 
-        if let Some(loc) = self.find_primary_location(ex) {
-            self.locations_visited.insert(loc);
-            let rws = self.rewrites.get(&loc).map_or(&[] as &[_], |v| v);
-            for rw in rws {
-                hir_rw = match rw.kind {
-                    mir_op::RewriteKind::OffsetSlice { mutbl } => {
-                        // `p.offset(i)` -> `&p[i as usize ..]`
-                        assert!(matches!(hir_rw, Rewrite::Identity));
-                        let arr = self.get_subexpr(ex, 0);
-                        let idx =
-                            Rewrite::Cast(Box::new(self.get_subexpr(ex, 1)), "usize".to_owned());
-                        let elem = Rewrite::SliceTail(Box::new(arr), Box::new(idx));
-                        Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl))
-                    }
-
-                    mir_op::RewriteKind::SliceFirst { mutbl } => {
-                        // `p` -> `&p[0]`
-                        let arr = hir_rw;
-                        let idx = Rewrite::LitZero;
-                        let elem = Rewrite::Index(Box::new(arr), Box::new(idx));
-                        Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl))
-                    }
-
-                    mir_op::RewriteKind::MutToImm => {
-                        // `p` -> `&*p`
-                        let place = Rewrite::Deref(Box::new(hir_rw));
-                        Rewrite::Ref(Box::new(place), hir::Mutability::Not)
-                    }
-
-                    mir_op::RewriteKind::RemoveAsPtr => {
-                        // `slice.as_ptr()` -> `slice`
-                        assert!(matches!(hir_rw, Rewrite::Identity));
-                        self.get_subexpr(ex, 0)
-                    }
-
-                    mir_op::RewriteKind::RawToRef { mutbl } => {
-                        // &raw _ to &_ or &raw mut _ to &mut _
-                        Rewrite::Ref(Box::new(self.get_subexpr(ex, 0)), mutbl_from_bool(mutbl))
-                    }
-                };
-            }
+        let locs = self.find_locations(ex);
+        for loc in &locs {
+            self.locations_visited.insert(*loc);
+        }
+
+        let all_rws: Vec<&_> = locs
+            .iter()
+            .filter_map(|loc| self.rewrites.get(loc))
+            .flatten()
+            .collect();
```

Comment:
I'm not sure if I like this strategy.  My original thinking when writing `find_sole_location` was that for each HIR expr, we should find all the MIR statements it produced and label each one with its purpose/origin within the expr, since `mir_op` rewrites on different parts may need to be applied in different ways.  For `addr_of!` this will work, but I worry that lumping together all the rewrites will cause problems for more complex situations where a single expr might produce 3-4 different statements corresponding to different parts of the expression.

Reply:
is the `flatten` the crux of the issue then, with the unflattened version being the crudest possibility of separating groups of MIR statements resulting from HIR expressions?

Reply:
Would it be reasonable to group rewrites by the span of the HIR expression? Something analogous to a rewrite tree per HIR expression? I'm not sure how else to support multiple MIR statements per HIR Assign expression, for example. What do you recommend?

Reply:
There's no general strategy for this.  Essentially what we're doing here is reverse-engineering parts of rustc's HIR-to-MIR lowering so that we can lift rewrites from the MIR to the HIR.  I think the safest way to have separate cases for all the specific patterns we know how to handle.  For `addr_of!` specifically, the pattern is something like: a HIR `ExprKind::AddrOf` whose span came from a macro expansion, and exactly two adjacent MIR statements with the same span where the first is an `Rvalue::Ref` assignment and the second is an `Rvalue::AddressOf` assignment.  And in this case, a `mir::RewriteKind::RawToRef` on the second MIR statement is lifted into a corresponding HIR-level `Rewrite::Ref`, while other kinds of rewrites on the second statement and any rewrites on the first statement are ignored / cause a panic / get some default handling (I'm not sure of the exact details here).

It's possible to make this stricter (e.g. we expect that the LHS of the first assignment is a temporary, which appears only in the RHS of the second assignment), which reduces the possibility of bad rewrites but also means more work for us in writing and adjusting ultra-specific patterns.  It's also possible to make it more lenient (e.g. the current approach where all rewrites are pooled together), but it seems likely to me that this will lead to bugs in the future around things like method calls, where the call itself and the deref adjustment on the receiver have the same span but may have two different sets of MIR rewrites that need to be handled in different ways.

Reply:
Yes, that looks pretty good.  We can rearrange the individual cases later if needed.